Fix #4532: Prevent multiple concept card fragments

masclot opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fix #4532: only allow one concept card open at a time. If the new new concept card has the same skill id as an existing one, do not open and instead reuse the existing one. 

I modified a couple of concept card data for dev so that they link to each other. This change is not necessary for this PR to work, but it allows manually testing the scenario where a concept car links to another.

Current behavior:
[multiple concept cards allowed, even with same skill id.webm](https://user-images.githubusercontent.com/103062089/232325925-d9ce7ad2-a260-46a8-a23a-075d0152f48e.webm)

New behavior:
[ConceptCard deduplication demo.webm](https://user-images.githubusercontent.com/103062089/232325566-96c74599-a5b8-4201-9feb-37f71c34649c.webm)

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

masclot committed:
Merge sw600 versions of profile_chooser_fragment.xml

masclot committed:
Merge branch 'oppia:develop' into merge-profile-chooser-xml

masclot committed:
Merge remote-tracking branch 'upstream/develop' into develop
Checking out the latest code in 'develop'

masclot committed:
Merge remote-tracking branch 'upstream/develop' into develop

masclot committed:
Merge remote-tracking branch 'upstream/develop' into develop
Update repository

masclot committed:
Do not open a child fragment if one with the same tag already exists. This prevents opening two dialog fragments on double click.

masclot committed:
Fix line length

masclot committed:
Fix formatting issues

BenHenning commented:
@masclot is this ready for review or did you open the draft so that people can help with specific parts? Just wanted to make sure that we aren't inadvertently ignoring this. :)

masclot commented:
> @masclot is this ready for review or did you open the draft so that people can help with specific parts? Just wanted to make sure that we aren't inadvertently ignoring this. :)

It is not ready. Since the change is not trivial, I just wanted to show what my idea is, so that I get early comments. @adhiamboperes already commented on chat.

masclot committed:
Add support for multiple, simultaneous concept cards, each with its own tag. Allow to switch concept cards by reusing instances instead of creating a new one each time.

masclot committed:
Fix formatting.

masclot committed:
Two things:
- Modify two test concept cards to link to each other; this allows testing the functionality
- Only allow one concept card at a time

masclot committed:
Two things:
- Finish implementation
- Add unit tests

masclot committed:
Fix formatting

masclot commented:
@adhiamboperes PTAL

masclot commented:
@BenHenning PTAL

masclot committed:
Fix static check: use InjectableDialogFragment instead of DialogFragment

masclot commented:
The error in the checks seems to be temporal:

\# A fatal error has been detected by the Java Runtime Environment:
\#
\#  SIGSEGV (0xb) at pc=0x00007f02ba37d3e0, pid=2671, tid=2685

Could anybody rerun them?

adhiamboperes commented:
Thanks @masclot! I took a first pass, and I'll be able to provide more specific feedback in the next pass.

-  Since you've made changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files , please have their rationale included in the PR explanation.
- There are some unrelated commits/changes, please remove them.

adhiamboperes reviewed (changes requested):
Thanks @masclot! I took a first pass, and I'll be able to provide more specific feedback in the next pass.

-  Since you've made changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files , please have their rationale included in the PR explanation.
- There are some unrelated commits/changes, please remove them.

BenHenning commented:
Hi @masclot. Apologies, will need to take a look at this tomorrow.

Also updating this to the latest develop branch.

BenHenning committed:
Merge branch 'develop' into fix_double_click_in_card

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/ConceptCardFragmentTestActivity.kt:
```diff
@@ -31,15 +31,15 @@ class ConceptCardFragmentTestActivity : InjectableAppCompatActivity(), ConceptCa
 
   private fun getConceptCardFragment(): ConceptCardFragment? {
-    return supportFragmentManager.findFragmentByTag(TAG_CONCEPT_CARD_DIALOG) as ConceptCardFragment?
+    return supportFragmentManager.fragments.singleOrNull { fragment ->
+      fragment is ConceptCardFragment
+    } as ConceptCardFragment?
```
```suggestion
    return supportFragmentManager.fragments.filterIsInstance<ConceptCardFragment>().single()
```

Bit more robust and equivalent.

masclot replied:
Done, but kept singleOrNull() instead of single() to keep the original method signature.

BenHenning replied:
Ah, I think I meant ``singleOrNull`` originally (per your existing code)--good catch.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragment.kt:
```diff
@@ -22,23 +23,74 @@ class ConceptCardFragment : InjectableDialogFragment() {
+     * skill Id. If no [ConceptCardFragment] remains, creates a new fragment to show a concept card
+     * for the given skill id.
+     *
+     * @param skillId the skill ID for which a concept card should be loaded.
+     * @param profileId the profile in which the concept card will be shown.
+     * @param fragmentManager the [FragmentManager] where to show the concept card.
+     * @return the fragment created or null if none.
```
```suggestion
     * @param skillId the skill ID for which a concept card should be loaded
     * @param profileId the profile in which the concept card will be shown
     * @param fragmentManager the [FragmentManager] where to show the concept card
     * @return the fragment created or null if none
```

Small nit on punctuation consistency for KDoc params.

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragment.kt:
```diff
@@ -22,23 +23,74 @@ class ConceptCardFragment : InjectableDialogFragment() {
+      profileId: ProfileId,
+      fragmentManager: FragmentManager
+    ): ConceptCardFragment? {
+      // The UI is recreated when changing profile ids, so no need to include it in the tag name.
```
This comment seems a bit out-of-band--perhaps add more context to explain the potential issue & why this means it isn't a problem?

masclot replied:
Expanded the comment. I could also complicate the code so that the comment is not necessary.

BenHenning replied:
I think the new comment is much clearer now--thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragment.kt:
```diff
@@ -75,4 +127,6 @@ class ConceptCardFragment : InjectableDialogFragment() {
     dialog?.window?.setWindowAnimations(R.style.FullScreenDialogStyle)
   }
+
+  private fun skillId(): String? = arguments?.getStringFromBundle(SKILL_ID_ARGUMENT_KEY)
```
nit: ``getSkillId`` per codebase consistency.

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragment.kt:
```diff
@@ -22,23 +23,74 @@ class ConceptCardFragment : InjectableDialogFragment() {
+        }
+        transaction.commitNow()
+      }
+      if (!found) {
+        return showNewInstance(skillId, profileId, fragmentManager)
+      }
+      return null
```
```suggestion
      return if (allConceptCards.size <= conceptCardsWithDifferentSkillId.size) {
        showNewInstance(skillId, profileId, fragmentManager)
      } else null
```

Some minor simplifications.

masclot replied:
Done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragmentTest.kt:
```diff
@@ -363,6 +370,112 @@ class ConceptCardFragmentTest {
   }
 
+  @Test
+  fun testConceptCardFragment_launchSeveralConceptCardsWithSameSkill_onlyTheFirstOne() {
```
```suggestion
  fun testConceptCardFragment_launchSeveralConceptCardsWithSameSkill_onlyTheFirstShows() {
```

masclot replied:
Done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragmentTest.kt:
```diff
@@ -363,6 +370,112 @@ class ConceptCardFragmentTest {
+          ProfileId.getDefaultInstance(),
+          activity.supportFragmentManager
+        )
+        Assert.assertNotNull(fragment)
+        Assert.assertNull(
+          ConceptCardFragment.bringToFrontOrCreateIfNew(
+            TEST_SKILL_ID_0,
+            ProfileId.getDefaultInstance(),
+            activity.supportFragmentManager
+          )
+        )
+        Assert.assertEquals(1, activity.supportFragmentManager.fragments.size)
+        Assert.assertEquals(fragment, activity.supportFragmentManager.fragments[0])
```
Prefer using Truth assertions over JUnit's (and statically import them). Ditto for the other tests.

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragment.kt:
```diff
@@ -22,23 +23,74 @@ class ConceptCardFragment : InjectableDialogFragment() {
+      skillId: String,
+      profileId: ProfileId,
+      fragmentManager: FragmentManager
+    ): ConceptCardFragment? {
```
Is the return actually useful? It seems like none of the callsites use it (unless I missed one of them).

masclot replied:
I use the return in tests.

BenHenning replied:
Having an API specifically for tests isn't generally preferred unless that API may be used by non-tests in the future, or is test-only (since it sort of skews the expected usage of the API between production and test contexts).

If the goal here is just to verify that the correct outcome happens, could we instead inspect the activity's fragment manager to see if the correct fragment has been created/not created rather than depending on the return value of this function? I think that will result in both more robust tests and a cleaner API.

masclot replied:
Done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragmentTest.kt:
```diff
@@ -363,6 +370,112 @@ class ConceptCardFragmentTest {
+  }
+
+  @Test
+  fun testConceptCardFragment_severalConceptCards_dismissAll() {
```
Perhaps add a test for dismissAll when there are no cards?

masclot replied:
Done

BenHenning reviewed (changes requested):
Thanks @masclot! Apologies for the delayed review.

I think this solution is fantastic. It's much more elegant and UI-isolated than I was originally thinking, so I'm quite happy with what you've come up with here. Mostly had style nits--PTAL.

masclot committed:
Address PR comments

masclot commented:
> Thanks @masclot! I took a first pass, and I'll be able to provide more specific feedback in the next pass.
> 
> * Since you've made changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files , please have their rationale included in the PR explanation.
> * There are some unrelated commits/changes, please remove them.

@adhiamboperes 
I added a comment about the asset changes  to the PR description.
I couldn't find unrelated commits/changes. Could you point me to them?

masclot commented:
@BenHenning 
@adhiamboperes 
PTAL

adhiamboperes reviewed (approved):
I've looked over your changes and they look really good!

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragmentTest.kt:
```diff
@@ -371,24 +371,24 @@ class ConceptCardFragmentTest {
-        Assert.assertEquals(1, activity.supportFragmentManager.fragments.size)
-        Assert.assertEquals(fragment, activity.supportFragmentManager.fragments[0])
+        ).isNull()
+        assertThat(activity.supportFragmentManager.fragments.size).isEqualTo(1)
```
```suggestion
        assertThat(activity.supportFragmentManager.fragments).hasSize(1)
```

More Truth-idiomatic version. Ditto elsewhere.

masclot replied:
Done

BenHenning reviewed (commented):
Thanks @masclot! Just had a few follow-ups--PTAL.

masclot committed:
Address PR comments

masclot commented:
Addressed the last two conversations.
@BenHenning PTAL

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/conceptcard/ConceptCardFragmentTest.kt:
```diff
@@ -420,23 +422,25 @@ class ConceptCardFragmentTest {
+        assertThat(activity.supportFragmentManager.fragments).hasSize(2)
         ConceptCardFragment.dismissAll(activity.supportFragmentManager)
-        assertThat(activity.supportFragmentManager.fragments.size).isEqualTo(0)
+        assertThat(activity.supportFragmentManager.fragments).hasSize(0)
```
```suggestion
        assertThat(activity.supportFragmentManager.fragments).isEmpty()
```

Small nit: ``isEmpty()`` is the more idiomatic method for this case.

masclot replied:
Done

BenHenning reviewed (commented):
Thanks @masclot! Really sorry for the delayed review here.

The PR LGTM, but had 1 nit left and I think the branch needs to be brought up-to-date with the latest develop changes. Please assign back once both are done and I'll go ahead and kick-off the merge.

masclot committed:
Address PR comments

masclot commented:
> Thanks @masclot! Really sorry for the delayed review here.
> 
> The PR LGTM, but had 1 nit left and I think the branch needs to be brought up-to-date with the latest develop changes. Please assign back once both are done and I'll go ahead and kick-off the merge.

I addressed the comment. Regarding "bringing the branch up-to-date with the latest develop changes", I'm not sure I'm doing it right. In Android Studio, I checked out my develop branch and clicked "Update" on it. I then checked-out the branch for this change and also clicked "Update". Would that be enough?

masclot committed:
Merge remote-tracking branch 'upstream/develop' into develop

masclot committed:
Merge branch 'develop' into fix_double_click_in_card

masclot commented:
I don't think Android Studio -> \<branch-name\> -> Update does a merge.  This menu is what oppia documentation says we should do in Android Studio to update our repository: https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#making-a-local-code-change-using-android-studios-ui-based-github-workflow

Instead I did the following in Android Studio:
- checkout develop
- fetch
- merge upsteam/develop
- checkout \<my-branch\>
- fetch
- merge develop

I hope these steps are good. Please let me know otherwise.

BenHenning reviewed (approved):
Thanks @masclot! Re: updating the branch, I think we have instructions in the wiki for this but I can't find them off hand at the moment (@MohitGupta121 might you know?). We generally recommend doing this via Git CLI since it's easier to "debug" when something goes wrong, and I'm less familiar with the Android Studio Git interface.

PR LGTM!

BenHenning commented:
Updating to latest develop & enabling auto-merge.

BenHenning committed:
Merge branch 'develop' into fix_double_click_in_card

adhiamboperes committed:
Merge branch 'develop' into fix_double_click_in_card
